src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -894,72 +893,59 @@ static BOOL AttributeHasChangesOmittedQuality(MTRAttributePath * attributePath)
             NSMutableArray<NSArray *> * readRequestsCurrent = opaqueDataCurrent;
             NSMutableArray<NSArray *> * readRequestsNext = opaqueDataNext;
 
-            // Can only read up to 9 paths at a time, per spec
-            if (readRequestsCurrent.count >= 9) {
-                MTR_LOG_DEFAULT("%@ batching cannot add more", logPrefix);
-                return;
-            }
-
+            MTRBatchingOutcome outcome = MTRNotBatched;
             while (readRequestsNext.count) {
+                // Can only read up to 9 paths at a time, per spec
+                if (readRequestsCurrent.count >= 9) {
+                    MTR_LOG_INFO("Batching cannot add more work, current item is full");
+                    return outcome;
+                }
+
                 // if params don't match then they cannot be merged
                 if (![readRequestsNext[0][MTRDeviceReadRequestFieldParamsIndex]
                         isEqual:readRequestsCurrent[0][MTRDeviceReadRequestFieldParamsIndex]]) {
-                    MTR_LOG_DEFAULT("%@ batching merged all possible items", logPrefix);
-                    return;
+                    MTR_LOG_INFO("Batching cannot add more work, parameter mismatch");
+                    return outcome;
                 }
 
                 // merge the next item's first request into the current item's list
+                MTR_LOG_INFO("Batching %@ => %tu total", readRequestsNext[0], readRequestsCurrent.count);
                 [readRequestsCurrent addObject:readRequestsNext[0]];
-                MTR_LOG_INFO("%@ batching merging %@ => %lu total", logPrefix, readRequestsNext[0],
-                    (unsigned long) readRequestsCurrent.count);
                 [readRequestsNext removeObjectAtIndex:0];
-
-                // Can only read up to 9 paths at a time, per spec
-                if (readRequestsCurrent.count == 9) {
-                    MTR_LOG_DEFAULT("%@ batching to max paths allowed", logPrefix);
-                    break;
-                }
-            }
-
-            if (readRequestsNext.count == 0) {
-                MTR_LOG_DEFAULT("%@ batching - fully merged next item", logPrefix);
-                *fullyMerged = YES;
+                outcome = MTRBatchedPartially;
             }
+            NSCAssert(readRequestsNext.count == 0, @"should have batched everything or returned early");
+            return MTRBatchedFully;
         }];
         [workItem setDuplicateTypeID:MTRDeviceWorkItemDuplicateReadTypeID handler:^(id opaqueItemData, BOOL * isDuplicate, BOOL * stop) {
+            mtr_hide(self); // don't capture self accidentally
             for (NSArray * readItem in readRequests) {
                 if ([readItem isEqual:opaqueItemData]) {
-                    MTR_LOG_DEFAULT("%@ duplicate check found %@ - report duplicate", logPrefix, readItem);
+                    MTR_LOG_DEFAULT("Read attribute duplicate check found %@ - report duplicate", readItem);
```

Comment:
Again, might be nice to know the id of the item that caused the duplicate to be dropped, unless the queue logs it.